developer/src/tike/oskbuilder/UframeTouchLayoutBuilder.pas

```diff
@@ -332,71 +334,85 @@ function TframeTouchLayoutBuilder.Load(const AFilename: string; ALoadFromTemplat
   end;
 
   UnregisterSources;
-  try
 
-    if ALoadFromString then
+  if ALoadFromString then
+  begin
+    FNewLayoutJS := AFilename;
+    FFilename := GetNextFilename;
+  end
+  else
+  begin
+    if ALoadFromTemplate or (AFileName = '') or not FileExists(AFileName) then
     begin
-      FNewLayoutJS := AFilename;
+      FBaseFileName := FTemplateFileName;
       FFilename := GetNextFilename;
     end
     else
     begin
-      if ALoadFromTemplate or (AFileName = '') or not FileExists(AFileName) then
-      begin
-        FBaseFileName := FTemplateFileName;
-        FFilename := GetNextFilename;
-      end
-      else
-      begin
-        FBaseFileName := AFileName;
-        FFilename := AFileName;
-      end;
-
-      with TStringList.Create do
-      try
-        LoadFromFile(FBaseFileName, TEncoding.UTF8);
-        FNewLayoutJS := Text;
-      finally
-        Free;
-      end;
+      FBaseFileName := AFileName;
+      FFilename := AFileName;
     end;
 
-    FTouchLayout := TTouchLayout.Create;   // I3642
+    with TStringList.Create do
     try
-      if not FTouchLayout.Load(FNewLayoutJS) then
+      LoadFromFile(FBaseFileName, TEncoding.UTF8);
+      FNewLayoutJS := Text;
+    finally
+      Free;
+    end;
+  end;
+
+  FTouchLayout := TTouchLayout.Create;   // I3642
+  try
+    if not FTouchLayout.Load(FNewLayoutJS) then
+    begin
+      FLastError := FTouchLayout.LoadError;   // I4083
+      FLastErrorOffset := FTouchLayout.LoadErrorOffset;   // I4083
+      FFilename := FLastFilename;
+      RegisterSources(FState);
```

Comment:
Because I eliminated the `try`/`finally` pattern, we have to call RegisterSources here to unwind the earlier UnregisterSources from line 336.